libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp

```diff
@@ -0,0 +1,449 @@
+    _dev->register_periodic_callback(1000, FUNCTOR_BIND_MEMBER(&AP_InertialSensor_LSM6DS3::_poll_data, void));
+}
+
+uint8_t AP_InertialSensor_LSM6DS3::_register_read(uint8_t reg)
+{
+    uint8_t val = 0;
+    _dev->read_registers(reg, &val, 1);
+    return val;
+}
+
+
+void AP_InertialSensor_LSM6DS3::_register_write(uint8_t reg, uint8_t val, bool checked)
+{
+    _dev->write_register(reg, val, checked);
+}
+
+void AP_InertialSensor_LSM6DS3::_gyro_init()
+{
+    _register_write(LSM6DS3_CTRL2_G, LSM6DS3_CTRL2_G_ODR_833Hz |
+                                              LSM6DS3_CTRL2_G_FS_2000DPS);
+    hal.scheduler->delay(1);
+
+    _set_gyro_scale();
+    hal.scheduler->delay(1);
+}
+
+void AP_InertialSensor_LSM6DS3::_accel_init()
+{
+    _register_write(LSM6DS3_CTRL1_XL, LSM6DS3_CTRL1_XL_ODR_833Hz |
+                                               LSM6DS3_CTRL1_XL_FS_16G);
+    hal.scheduler->delay(1);
+
+    _set_accel_scale(A_SCALE_16G);
+    hal.scheduler->delay(1);
+}
+
+void AP_InertialSensor_LSM6DS3::_set_gyro_scale()
+{
+    /* scale value from datasheet 2000 mdps/digit */
+    _gyro_scale = 70;
+    /* convert mdps/digit to dps/digit */
+    _gyro_scale /= 1000;
+    /* convert dps/digit to (rad/s)/digit */
+    _gyro_scale *= DEG_TO_RAD;
+}
+
+void AP_InertialSensor_LSM6DS3::_set_accel_scale(accel_scale scale)
+{
+    /*
+     * Possible accelerometer scales (and their register bit settings) are:
+     * 2 g (000), 4g (001), 6g (010) 8g (011), 16g (100). Here's a bit of an
+     * algorithm to calculate g/(ADC tick) based on that 3-bit value:
+     */
+    _accel_scale = (((float) scale + 1.0f) * 2.0f) / 32768.0f;
+    if (scale == A_SCALE_16G) {
+        /* the datasheet shows an exception for +-16G */
+        _accel_scale = 0.000488f;
+    }
+    /* convert to G/LSB to (m/s/s)/LSB */
+    _accel_scale *= GRAVITY_MSS;
+}
+
+/**
+ * Timer process to poll for new data from the LSM6DS3.
+ */
+void AP_InertialSensor_LSM6DS3::_poll_data()
+{
+    uint16_t samples=0;
+
+    uint16_t frames = _register_read(LSM6DS3_FIFO_STATUS1);
+     
+    frames = frames & LSM6DS3_FIFO_STATUS1_UNREAD_WORDS;
+    
+    //(GYRO + XL) = 12 bytes (6 words)
+    samples = frames / 6;
+
+    if (samples > 1) {
+    	_read_data_transaction(samples);    
+    }
+
+    if (frames == 30) {
+        _fifo_reset();
+        //hal.console->printf("FIFO_RESET\n");
+    }
+
+    // check next register value for correctness
+    AP_HAL::Device::checkreg reg;
+    if (!_dev->check_next_register(reg)) {
+        log_register_change(_dev->get_bus_id(), reg);
+        _inc_accel_error_count(_accel_instance);
+    }
+
+}
+
+
+/*
+ *  read from the FIFO data output registers
+ */
+void AP_InertialSensor_LSM6DS3::_read_data_transaction(uint16_t samples)
+{
+    struct sensor_data fifo_data_gyro = { };
+    struct sensor_data fifo_data_accel = { };
+    
+    int32_t _gyro_bias[3] = {0, 0, 0};
+    int32_t _accel_bias[3] = {0, 0, 0};
```

Comment:
```suggestion
    struct sensor_data fifo_data_gyro { };
    struct sensor_data fifo_data_accel { };
    
    int32_t _gyro_bias[3] {};
    int32_t _accel_bias[3] {};
```

Save some bytes; stop creating objects on stack and assigning them back into other object